lib/spack/spack/schema/container.py

```diff
@@ -4,6 +4,43 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 """Schema for the 'container' subsection of Spack environments."""
 
+_stages_from_dockerhub = {
+    'type': 'object',
+    'additionalProperties': False,
+    'properties': {
+        'os': {
```

Comment:
(main concern) If I understand correctly, "build" and "final" are used in the same way as "os" and "spack". They have different names to allow you to distinguish the constrained and unconstrained case in the schema. I'm inclined to say it would be clearer to:

* add a separate section which lists valid build/final images when a user wants to constrain them
  * ... and add these values to the default config (in `etc/spack`)
* add a variable that determines whether Spack should ensure that a specified build/final image is a preapproved one
  * ... and default that variable to true

That way the user doesn't have to choose different config variables to achieve the same effect.

Reply:
> If I understand correctly, "build" and "final" are used in the same way as "os" and "spack". They have different names to allow you to distinguish the constrained and unconstrained case in the schema. I'm inclined to say it would be clearer to [ ... ]

I don't really agree with this. The "os" + "spack" configuration is basically a simplified way to configure image generation if a user is fine with starting from our base images on Dockerhub. This is achieved by:

1. Constraining the images (as you noted)
2. Ensuring consistency between base images in the build and final stage

Consider for instance:
```yaml
images:
  os: ubuntu:18.04
  spack: 0.14.1
```
compared to what is my understanding of the proposal above:
```yaml
constrained: True
images:
  build: spack/ubuntu-bionic:0.14.1
  final: ubuntu:18.04
```
In the second case we require to the user to be needlessly more verbose and duplicate information (`ubuntu-bionic` and `ubuntu:18.04`), in the first case we just ask the minimal set of information we need to proceed with generating the container recipe. 

Overall I don't think having something in-between what we have now and removing the simplified mode is worthwhile, as:

1. We wouldn't be that much of an help in simplifying configuration to the user
2. We need to maintain nonetheless all the code that enforces constraints

My personal opinion would be to maintain this as is, but if people feel strongly I would be for just removing the "os" + "spack" mode and have no "simplified mode".

> I have a few minor requests but overall I think that "build"/"final" and "os"/"spack" are duplicate ways to specify the same thing, which I think is confusing.

Can you elaborate on what can be confusing / what do you think will annoy users with this setup? Asking because the discussion here involves a few power users and this is the first time the argument is raised, so I'd like to understand better the concerns.

Reply:
> 1. We wouldn't be that much of an help in simplifying configuration to the user
> Asking because the discussion here involves a few power users

Introducing two different ways of specifying the same thing is confusing to regular users. The variables have overlapping meaning but change behavior based on which ones you choose. That's easy to mix up.

> 2. We need to maintain nonetheless all the code that enforces constraints

It took me a while to realize how much these variables overlap when I read through the code, so I also think this would improve readability. 

I think we should try to figure out how to maintain a single set of config variables to specify containers and then add logic to improve things for power users. For example:

```
constrained: True
images:
  build: ubuntu:18.04
  final: 0.14.1
```

Is there any way Spack could infer the values of `spack/ubuntu-bionic:0.14.1`/`ubuntu:18.04` from `ubuntu:18.04`/`0.14.1`?

The `constrained` config variable could also move "up" and apply simultaneously to all specs to reduce tedium (although in that case it should have a more-descriptive name).

Reply:
> Introducing two different ways of specifying the same thing is confusing to regular users. The variables have overlapping meaning but change behavior based on which ones you choose. That's easy to mix up.

I tried to solve this issue by adding explicit docs. See 63333ba530d6c8a72ab7578a795f75ba4e1c593e. Maybe that should be better worded?

> I think we should try to figure out how to maintain a single set of config variables to specify containers and then add logic to improve things for power users.

What we might do is the opposite of your example + avoid the `constrained` variable:
```yaml
images:
  build: 0.14.1
  final: ubuntu:18.04
```
Here the idea is that if `build` has a certain format it is interpreted as a Spack version, otherwise it's interpreted as an image on Docker hub. Personally I'd still prefer to have different labels since they make explicit this change, but let me know if that can be a good compromise.

> It took me a while to realize how much these variables overlap when I read through the code, so I also think this would improve readability.

Hmm, I don't see that in the sense that inner logic would remain the same but exposed to user in a different way. Maybe I just don't understand what you mean here.

Reply:
@scheibelp I got some additional feedback for the current interface from a user (reporting it here to add on the discussion):
1. The simplified interface with `os` and `spack` is seen as intuitive (i.e. understandable from just reading the YAML) so those names should not be abandoned
2. The interface with `build` and `final` is more confusing than the simplified interface and it's difficult to understand the meaning of those attributes if you don't read the documentation carefully
3. The proposal to maintain only `build` and `final` (or only `os` and `spack`) and switch their meaning with a Boolean is equally confusing

The proposal that was raised after a bit of discussion is to:
- Maintain the two sets of variables ("`os`+`spack`" and "`build` + `final`") with the meaning they have now
- Add a Boolean variable under `images` named `simplified` that defaults to `True` and controls whether we want to use one set of variables or the other.

In that way the simplified interface will look the same as it is now, but the switch of meaning requires an explicit step by the user. For instance:
```yaml
images:
  os: ubuntu:18.04
  spack: 0.14.1
```
would be an example of the simplified interface, while
```yaml
images:
  simplified: False
  build: alalazo/cuda-10.1-ubuntu18.04:latest
  final: nvidia/cuda:10.1-base-ubuntu18.04
```
would be how custom images can be chosen.

Reply:
Would the proposal above to have an explicit switch for the interface be fine with you?

Reply:
I think we agreed that if more users think these separate configuration variables are useful (i.e. os/spack for the pre-made case and build/final for the custom) case, then that's what we support. If you can find folks to weigh in on it then we can make progress on that; if not then we can keep discussing it.

If we end up maintaining both os/spack and build/final, I think there needs to be some work to improve how they are handled in the code to make it obvious to future maintainers what they are intended for.

Reply:
@hartzell @victorusu @paulbry Wondering if you have an opinion on the discussion above. The main point is if you find confusing the current interface with os/spack and a build/final to select between the two ways to specify images and if you have suggestions on how it should look like from user's perspective.

Reply:
Pinging also @nazavode since I already bothered him for https://github.com/spack/spack/pull/15028#discussion_r403925148

Reply:
I think @alalazo summarized well my points about the UI in https://github.com/spack/spack/pull/15028#discussion_r403925148. TL;DR, as a user:

1. _simplified UI_ should be the default, be intuitive, work out of the box, straightforward;
2. _advanced UI_ should be inconvenient, require user intervention and mutually exclusive.

The proposed boolean switch checks all the marks imo, I could reach out to some users I'm currently working with to ask for their impressions if could be beneficial.

Reply:
@scheibelp Anything I can do to help this discussion on the YAML format moving forward? Is https://github.com/spack/spack/pull/15028#discussion_r403925148 fine? Should we seek more opinions on Slack?

Reply:
At the risk of suggesting something that's none of the above, perhaps we could require that one of two section be specified, either

```yaml
stock_images:
  os: ubuntu:18.04
  spack: 0.14.1
```

or 

```yaml
custom_images:
  build: my_build_image
  final: my_final_image
```

It took me a while to figure out that the `spack` key was a [commit-ish](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefcommit-ishacommit-ishalsocommittish) within the Spack repository, presumably that'll be documented (assuming of course that a tag/branch/commit is valid there.

Reply:
@hartzell There's doc already for that in this PR, see [here](https://github.com/spack/spack/blob/d6acdd20ffe213e48e13d017b77a21058e18931f/lib/spack/docs/containers.rst). The part that explains possible values for `os` and `spack` is:

> Any combination of OS and Spack version that can be matched to an image in the Spack repository on Dockerhub is admissible as a value. By employing this mode the user instructs Spack to generate a recipe that will start from one of the official Spack images in the build stage and from the base OS image in the final stag

Reply:
Perhaps you could clarify this further with an example:

> For example to use the "spack/ubuntu-bionic:0.14" image, you can set...